edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs

```diff
@@ -77,72 +89,149 @@ public static (IList<X509Certificate2>, Option<string>) BuildCertificateList(X50
+                    string thumbprint = GetSha256Thumbprint(chainElement);
+                    if (remoteCertificateChain.Any(cert => GetSha256Thumbprint(cert) == thumbprint) &&
+                        caList.Any(cert => GetSha256Thumbprint(cert) == thumbprint))
+                    {
+                        match = true;
+                        break;
+                    }
+                }
+                if (!match)
+                {
+                    return (false, Option.Some($"Error validating cert with Subject: {remoteCertificate.SubjectName} Thumbprint: {GetSha256Thumbprint(remoteCertificate)}"));
+                }
+                else
                 {
-                    match = true;
-                    break;
+                    return (true, Option.None<string>());
                 }
             }
 
-            if (!match)
-            {
-                return (false, Option.Some($"Error validating cert with Subject: {remoteCertificate.SubjectName} Thumbprint: {GetSha256Thumbprint(remoteCertificate)}"));
-            }
-            else
+            return (true, Option.None<string>());
+        }
+
+        public static string GetCommonNameFromSubject(string subject)
+        {
+            string commonName = null;
+            string[] parts = subject.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
+
+            foreach (var part in parts)
             {
-                return (true, Option.None<string>());
+                var partTrimed = part.Trim();
+                if (partTrimed.StartsWith("CN", StringComparison.OrdinalIgnoreCase))
+                {
+                    string[] cnParts = partTrimed.Split(new char[] { '=' }, StringSplitOptions.RemoveEmptyEntries);
+                    if (cnParts.Length > 1)
+                    {
+                        commonName = cnParts[1].Trim();
+                    }
+                }
             }
+
+            return commonName;
+        }
+
+        public static bool ValidateCommonName(X509Certificate2 certificate, string commonName)
+        {
+            Preconditions.CheckNotNull(certificate);
+            Preconditions.CheckNotNull(commonName);
+
+            return String.Equals(commonName, GetCommonNameFromSubject(certificate.Subject), StringComparison.Ordinal);
+        }
+
+        public static bool ValidateCertificateThumbprint(X509Certificate2 certificate, IList<string> thumbprints)
+        {
+            Preconditions.CheckNotNull(certificate);
+            Preconditions.CheckNotNull(thumbprints);
+
+            return thumbprints.Any(th => th.ToLower() == certificate.Thumbprint.ToLower());
         }
 
-        public static bool ValidateClientCert(X509Certificate certificate, X509Chain chain, Option<IList<X509Certificate2>> caChainCerts, ILogger logger)
+        public static bool ValidateClientCert(X509Certificate certificate, X509Chain chain, Option<IList<X509Certificate2>> trustedCACerts, ILogger logger)
```

Comment:
Isn't the client cert always X509Certificate2 and chain a List<X509Certificate2>?

Reply:
Removed this redundant code. I kept it thinking this may be used for WS or AMQP.